eras/shelley/impl/src/Cardano/Ledger/Shelley/Delegation.hs

```diff
@@ -0,0 +1,352 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE ViewPatterns #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Cardano.Ledger.Shelley.Delegation (
+  ShelleyEraDCert (..),
+  pattern DCertMir,
+  pattern ShelleyDCertDeleg,
+  ShelleyDelegCert (..),
+  delegCWitness,
+  ShelleyDCert (..),
+  MIRCert (..),
+  MIRPot (..),
+  MIRTarget (..),
+  isRegKey,
+  isDeRegKey,
+  isDelegation,
+  isRegPool,
+  isRetirePool,
+  isGenesisDelegation,
+  isInstantaneousRewards,
+  isReservesMIRCert,
+  isTreasuryMIRCert,
+  requiresVKeyWitness,
+  shelleyDCertDecoder,
+
+  -- * Re-exports
+  EraDCert (..),
+  pattern DCertPool,
+  pattern DCertGenesis,
+  Delegation (..),
+  PoolCert (..),
+  poolCWitness,
+  ConstitutionalDelegCert (..),
+  genesisCWitness,
+)
+where
+
+import Cardano.Ledger.BaseTypes (invalidKey)
+import Cardano.Ledger.Binary (
+  DecCBOR (decCBOR),
+  DecCBORGroup (..),
+  Decoder,
+  EncCBOR (..),
+  EncCBORGroup (..),
+  FromCBOR (..),
+  ToCBOR (..),
+  TokenType (TypeMapLen, TypeMapLen64, TypeMapLenIndef),
+  decodeRecordNamed,
+  decodeRecordSum,
+  decodeWord,
+  encodeListLen,
+  listLenInt,
+  peekTokenType,
+ )
+import Cardano.Ledger.Coin (Coin (..), DeltaCoin)
+import Cardano.Ledger.Core
+import Cardano.Ledger.Credential (Credential (..), StakeCredential)
+import Cardano.Ledger.Crypto
+import Cardano.Ledger.Keys (KeyRole (..))
+import Cardano.Ledger.Shelley.Era (ShelleyEra)
+import Cardano.Ledger.Slot (EpochNo (..))
+import Control.DeepSeq (NFData (..), rwhnf)
+import Data.Map.Strict (Map)
+import Data.Maybe (isJust, isNothing)
+import Data.Word (Word8)
+import GHC.Generics (Generic)
+import NoThunks.Class (NoThunks (..))
+
+instance Crypto c => EraDCert (ShelleyEra c) where
+  {-# SPECIALIZE instance EraDCert (ShelleyEra StandardCrypto) #-}
+
+  type DCert (ShelleyEra c) = ShelleyDCert (ShelleyEra c)
+
+  mkDCertPool = ShelleyDCertPool
+
+  getDCertPool (ShelleyDCertPool c) = Just c
+  getDCertPool _ = Nothing
+
+  mkDCertGenesis = ShelleyDCertGenesis
+
+  getDCertGenesis (ShelleyDCertGenesis c) = Just c
+  getDCertGenesis _ = Nothing
+
+class EraDCert era => ShelleyEraDCert era where
+  mkDCertMir :: ProtVerAtMost era 8 => MIRCert (EraCrypto era) -> DCert era
+  getDCertMir :: DCert era -> Maybe (MIRCert (EraCrypto era))
```

Comment:
```suggestion
  getDCertMir :: ProtVerAtMost era 8 => DCert era -> Maybe (MIRCert (EraCrypto era))
```

This one also needs a constraint. We want to prevent any code that relies on `MIR` certificates to be used in Conway. This will have to be done in a later era regardless, so might as well. tackle this now. That being said, it will be better if we do this in a separate PR, this one is already pretty big